kic_module.c

```diff
@@ -0,0 +1,198 @@
+#include <linux/ktime.h>	// provides time related utilities
+
+#include "keyboard_interrupt_counter.h"
+
+MODULE_LICENSE("GPL");
+
+/*
+ * controlling exclusive access to the device via atomic operations
+ */
+enum {
+	CDEV_NOT_USED = 0,
+	CDEV_EXCLUSIVE_OPEN = 1,
+};
+static atomic_t already_open = ATOMIC_INIT(CDEV_NOT_USED);
+static struct class *kic_cls;
+
+static atomic64_t kic_interrupt_count = ATOMIC_INIT(0);	// count of interrupts
+static ktime_t kic_time = 0;		// time and date of last counter reset
+
+/*
+ * prototypes for purpose-specific functions
+ */
+static irqreturn_t kic_increment(int, void *);
+static void kic_reset(void);
+
+/*
+ * file operations
+ *
+ * maintains reference count, invoked by ioctl protocol
+ */
+static int kic_device_open(struct inode *inode, struct file *file)
+{
+	pr_info("KIC: Device Opened.\n");
+	try_module_get(THIS_MODULE);
+	return 0;
+}
+/*
+ * maintains reference count, invoked by ioctl protocol
+ */
+static int kic_device_release(struct inode *inode, struct file *file)
+{
+	pr_info("KIC: Device Freed.\n");
+	module_put(THIS_MODULE);
+	return 0;
+}
+/*
+ * maintains exclusive access to the device and provides communication with
+ * userspace
+ */
+static long kic_device_ioctl(struct file *file, unsigned int ioctl_mode,
+			     unsigned long ioctl_carry)
+{
+	if (atomic_cmpxchg(&already_open, CDEV_NOT_USED, CDEV_EXCLUSIVE_OPEN)) {
+		pr_alert("KIC: Device is busy.\n");
+		return -EBUSY;
+	}
+
+	pr_info("KIC: Activating IOCTL protocol.\n");
+	switch (ioctl_mode) {
+	case KIC_RESET_QUERY: {		// option for resetting counter
+		kic_reset();
+		pr_info("KIC: Counter has been reset.\n");
+		break;
+	}// case closed
+	case KIC_GET_COUNT_QUERY: {	// option for retrieving count
+		long long kic_carry = atomic64_read(&kic_interrupt_count);
+
+		if (copy_to_user((void __user *)ioctl_carry, (void *)&kic_carry,
+				 sizeof(long long)) != 0)
+				pr_alert("KIC: Copy to user failed.\n");
+			else
+				pr_info("KIC: Count retrieved at %lld.\n",
+					 kic_carry);
+		break;
+	}// case closed
+	case KIC_GET_TIME_QUERY: {	// option for retrieving time and date
+		if (kic_time == 0) {
+			pr_alert("KIC: Timer has not yet been reset.\n");
+			put_user(-1, (long int __user *)ioctl_carry);
+			break;
+		}
+		put_user(kic_time, (long int __user *)ioctl_carry);
+		pr_info("KIC: Time retrieved.\n");
+		break;
+	}// case closed
+	}// switch closed
+	atomic_set(&already_open, CDEV_NOT_USED);
+	return 0;
+}
+
+static struct file_operations fops = {
+	.owner = THIS_MODULE,
+	.open = kic_device_open,
+	.release = kic_device_release,
+	.unlocked_ioctl = kic_device_ioctl,
+};
+/*
+ * purpose specific functions
+ *
+ * increases count upon detecting a valid interrupt
+ */
+static irqreturn_t kic_increment(int irq, void *dev_id)
+{
+	atomic64_inc(&kic_interrupt_count);
+	return IRQ_HANDLED;
+}
+
+/*
+ * resets counter and saves time data when invoked
+ */
+static void kic_reset(void)
+{
+	kic_time = ktime_get_real();
```

Comment:
I would also initialize it in init.

Reply:
I've decided against it. Initializing it with date of loading makes sense, but then a custom message should be added to inform that this is the date of loading and not a reset itself. As it is, this is initialized with a 0 and used to print that particular info ("The counter has not yet been reset."). I don't see how either approach is inherently better so I'd rather leave it as it is.